internal/bull/upload.go

```diff
@@ -0,0 +1,67 @@
+package bull
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type uploadResponse struct {
+	File          string `json:"file"`          // Original filename
+	SavedFullPath string `json:"savedFullPath"` // Full path to the saved file
+	SavedFilename string `json:"savedFilename"` // Filename as saved
+}
+
+// upload Saves a file to <page>-<timestamp>-<original filename>
+func (b *bullServer) upload(w http.ResponseWriter, r *http.Request) error {
+	if b.editor == "" {
+		return httpError(http.StatusForbidden, fmt.Errorf("running in read-only mode (-editor= flag)"))
+	}
+
+	// Limit the size of the uploaded file to 25MB
+	err := r.ParseMultipartForm(25 << 20)
+	if err != nil {
+		return err
+	}
+
+	file, handler, err := r.FormFile("file")
+	if err != nil {
+		log.Printf("error reading file: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return err
+	}
+	defer file.Close()
+
+	pageName := pageFromURL(r)
+	savedFileName := fmt.Sprintf("%s-%d-%s", pageName, time.Now().UnixMilli(), handler.Filename)
+
+	if err := mkdirAll(b.content, filepath.Dir(savedFileName), 0755); err != nil {
+		return err
+	}
+	f, err := b.content.OpenFile(savedFileName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
```

Comment:
Instead of truncating (overwriting) existing files, why don’t we retry in a loop? Eventually, time.Now().UnixMilli() will change and result in a filename that isn’t taken.